model/flow/transaction_result.go

```diff
@@ -52,3 +52,14 @@ type LightTransactionResult struct {
 	// ComputationUsed is amount of computation used while executing the transaction.
 	ComputationUsed uint64
 }
+
+type TransactionResultErrorMessage struct {
+	// TransactionID is the ID of the transaction this result error was emitted from.
+	TransactionID Identifier
+	// Index is the index of the transaction this result error was emitted from.
+	Index uint32
```

Comment:
Yes, both the `Index` and `TransactionID` fields are necessary:
- `Index` is essential because when storing error messages, we need to retain the original index to ensure it aligns accurately with the transaction result index. Using just an iterator here it won’t work, as not all transactions fail, which would lead to mismatches between the index in the transaction result and the index in the error message.
https://github.com/onflow/flow-go/blob/dd56f17bb7b5ebb1e6f5c00deaa987cff4a210d2/storage/badger/transaction_result_error_messages.go#L137-L147
- `TransactionID` is required for the `backend` to handle `LookupErrorMessagesByBlockID` calls. This allows us to return the error messages associated with each transaction by its transaction ID.
https://github.com/onflow/flow-go/blob/dd56f17bb7b5ebb1e6f5c00deaa987cff4a210d2/engine/access/rpc/backend/backend_transactions.go#L1117-L1121